Added the bit stream reader feature.

jeffqjiangNew opened the pull request:
- This PR adds the new bit stream reader feature to rocDecode library core. The goal is to reduce ROCm package dependencies on FFMPEG and enable essential testings without FFMPEG demuxer.
- The bit stream reader contains a few built-in stream file parsers, including elementary stream file parser and IVF container file parser.
- Currently the reader can parse AVC, HEVC and AV1 elementary stream files and AV1 IVF container files. More format support will be added in the future.

jeffqjiangNew committed:
* rocDecode/ES parser: Added elementary stream file parser for HEVC and AVC.

jeffqjiangNew committed:
* rocDecode/ES parser: Added elementary stream file parser for AV1. Also cleaned up the bitstream ring buffer code.

jeffqjiangNew committed:
* rocDecode/ES parser: Added the IVF container file parser for AV1. Also fixed a bug in fill ring buffer function.

jeffqjiangNew committed:
* rocDecode/ES file parder: Added supported stream type detection.
 - The stream type detection checks the unique syntax patterns of the stream type and calculate the likeliheed score. Based on the score, the most likely type is determined.
 - The current supported stream types are: AVC/HEVC/AV1 elementary streams, IVF AV1 streams.

jeffqjiangNew committed:
* rocDecode/ES file parser: Fixed an AVC decode regression due to a copy and paste error.

jeffqjiangNew committed:
* rocDecode/ES file parser: Added bit depth parsing for codec support check; Added stronger AV1 detection for IVF AV1 stream type.

jeffqjiangNew committed:
* rocDecode/ES file parser: Removed debugging logs.

jeffqjiangNew committed:
* rocDecode/ES file parser: Added exmaple code to use the built-in file parser.

rrawther commented on utils/videoFileParser/video_es_parser.cpp:
```diff
@@ -0,0 +1,1159 @@
+
+#include <string.h>
+#include "video_es_parser.h"
+#include "../../src/parser/hevc_defines.h"
```
as we discussed, we need to expose another api for this and I don't think it will fit within the existing parser api

jeffqjiangNew replied:
Will remove the reference to the internal headers. 

Note this is a different parser. Think of it as a simple demuxer, which parses the files of AVC/HEVC/AV1 elementary stream files and get the picture data, and it can also read IVF container files. It is in parallel with roc_video_dec, ffmpeg demuxer in utility. The interface is exposed the same way as roc_video_dec, and ffmpeg demuxer. It is another utility.

rrawther commented on samples/videoDecode/videodecode.cpp:
```diff
@@ -57,7 +58,8 @@ void ShowHelpAndExit(const char *option = NULL) {
     << "-seek_criteria - Demux seek criteria & value - optional; default - 0,0; "
     << "[0: no seek; 1: SEEK_CRITERIA_FRAME_NUM, frame number; 2: SEEK_CRITERIA_TIME_STAMP, frame number (time calculated internally)]" << std::endl
     << "-seek_mode - Seek to previous key frame or exact - optional; default - 0"
```
I think, it is better to add another sample called "video_es_decode" for this instead of clubbing it with videodecode sample

jeffqjiangNew replied:
Agreed. The current code in videodecode.cpp is just a demo sample. When we remove ffmpeg links, we will need a new sample.

rrawther reviewed (changes requested):
added some comments. We can discuss in our meeting

AryanSalmanpour reviewed (commented):
@jeffqjiangNew Here is an idea about how to expose the new ES parser as an independent API and a new sample to show its functionality:

1. Move your implementation of the ES parser somewhere inside of the src/parser inside the rocdedecode library, this way you would have access to the internal codec structures headers.

2. Expose an independent API in the api/rocparser.h for example `extern rocDecStatus ROCDECAPI rocDecParserGetPicData(uint8_t **p_pic_data, int *pic_size);` notice that this is going to be an independent API without needing to use RocdecVideoParser parser_handle

3. Add a new function into utils/rocvideodecode/roc_video_dec.cpp/roc_video_dec.h `int GetPicData(uint8_t **p_pic_data, int *pic_size);` and inside that function call the new independent ES parser API (`rocDecParserGetPicData`)

4. Create a new sample (e.g., videoDecodeElementryStream) and inside that sample use the utils/rocvideodecode functionality and the new GetPicData function to illustrate the new ES parser without using any FFMPEG functionality

rrawther commented:
> @jeffqjiangNew Here is an idea about how to expose the new ES parser as an independent API and a new sample to show its functionality:
> 
> 1. Move your implementation of the ES parser somewhere inside of the src/parser inside the rocdedecode library, this way you would have access to the internal codec structures headers.
> 2. Expose an independent API in the api/rocparser.h for example `extern rocDecStatus ROCDECAPI rocDecParserGetPicData(uint8_t **p_pic_data, int *pic_size);` notice that this is going to be an independent API without needing to use RocdecVideoParser parser_handle
> 3. Add a new function into utils/rocvideodecode/roc_video_dec.cpp/roc_video_dec.h `int GetPicData(uint8_t **p_pic_data, int *pic_size);` and inside that function call the new independent ES parser API (`rocDecParserGetPicData`)
> 4. Create a new sample (e.g., videoDecodeElementryStream) and inside that sample use the utils/rocvideodecode functionality and the new GetPicData function to illustrate the new ES parser without using any FFMPEG functionality

I agree we need to expose a new API. But it can be a new class and API instead of clubbing it with existing parser api. Current parser is specifically designed for parsing picture data and sending appropriate information to rocDecode decoding

kiritigowda committed:
Merge branch 'develop' into jj/es_file_parser

jeffqjiangNew committed:
Merge branch 'develop' into jj/es_file_parser

jeffqjiangNew committed:
* rocDecode/Bitstream reader: Renamed the elementary parser feature to bitstream reader and re-organized the code.
 - Moved the bitstream reader code to rocDecode core lib from utility.
 - Added bitstream reader interface in parallel with rocDecode parser and decoder interfaces.

jeffqjiangNew committed:
* rocDecode/Bitstream reader: Added sample to use bitstream reader, instead of FFMPEG demuxer, to get picture data. Also reverted the original sample app back to using FFMPEG demuxer only.

jeffqjiangNew committed:
Merge branch 'develop' into jj/es_file_parser

jeffqjiangNew committed:
* rocDecode/Bitstream reader: Renamed the new sample app.

jeffqjiangNew committed:
* rocDecode/Bitstream reader: FFMPEG dependency reduction.
 - Moved MD5 functions out of RocVideoDecoder utility class. This removed RocVideoDecoder's dependency on FFMPEG.
 - Added the new MD5 utility, which depends on FFMPEG lib. MD5 message digest generation is now performed in the MD5 utility.
 - Modified decode sampples that uses MD5 generation function.
 - Removed FFMPEG dependency from video decoder basic sample.

jeffqjiangNew committed:
* rocDecode/Bitstream reader: Added option to use bitstream reader to video decode sample and conformance test script. Added the missing destroy bitstream reader call in video decode basic sample.

jeffqjiangNew committed:
* rocDecode/Bitstream reader: Minor format change. No functional changes.

jeffqjiangNew committed:
* rocDecode/Bitstream reader: Added handling of unsupported stream file type by the bitstream reader to decode sample apps.

jeffqjiangNew committed:
* rocDecode/Bitstream reader: Fixed build errors of several samples.

AryanSalmanpour commented on src/bitStreamReader/es_reader.h:
```diff
@@ -0,0 +1,236 @@
+#define INIT_PIC_DATA_SIZE (2 * 1024 * 1024)
+
+enum {
+    Stream_Type_UnSupported = -1,
```
For enum names, we have been using camel case prefixed by the letter 'k', such as 'kStreamTypeUnsupported.'

jeffqjiangNew replied:
Changed to Google style as suggested.

AryanSalmanpour commented on src/bitStreamReader/roc_bs_reader_api.cpp:
```diff
@@ -0,0 +1,100 @@
+}
+
+rocDecStatus ROCDECAPI rocDecGetBitstreamCodecType(RocdecBitstreamReader bs_reader_handle, rocDecVideoCodec *codec_type) {
+    if (bs_reader_handle == nullptr) {
```
Please include a check to ensure that the `codec_type` pointer is not a nullptr in this section.

jeffqjiangNew replied:
Done.

AryanSalmanpour commented on src/bitStreamReader/roc_bs_reader_api.cpp:
```diff
@@ -0,0 +1,100 @@
+}
+
+rocDecStatus ROCDECAPI rocDecGetBitstreamBitDepth(RocdecBitstreamReader bs_reader_handle, int *bit_depth) {
+    if (bs_reader_handle == nullptr) {
```
Please include a check to ensure that the `bit_depth` pointer is not a nullptr in this section.

jeffqjiangNew replied:
Done.

AryanSalmanpour commented on src/bitStreamReader/roc_bs_reader_api.cpp:
```diff
@@ -0,0 +1,100 @@
+}
+
+rocDecStatus ROCDECAPI rocDecGetBitstreamPicData(RocdecBitstreamReader bs_reader_handle, uint8_t **pic_data, int *pic_size, int64_t *pts) {
+    if (bs_reader_handle == nullptr || pic_data == nullptr) {
```
Please include a check to ensure that the `pic_size` and `pts` pointers are not a nullptr in this section.

jeffqjiangNew replied:
Done.

AryanSalmanpour commented on src/bitStreamReader/es_reader.cpp:
```diff
@@ -0,0 +1,1159 @@
+#include "roc_video_parser.h"
+
+RocVideoESParser::RocVideoESParser(char *input_file_path) {
+    p_stream_file_ = fopen(input_file_path, "rb");
```
Please use `std::ifstream` in C++ for all file input and output operations instead of using `fopen` and `fread`.

jeffqjiangNew replied:
Done. But 2 AV1 conformance streams are failing. Looking into it now.

jeffqjiangNew replied:
Done.

AryanSalmanpour reviewed (changes requested):
Please also rename the `bitStreamReader` folder to `bit_stream_reader` to match with the other folder names (e.g., amd_detail, decode, parser) under the `src` folder.

AryanSalmanpour commented on src/bitStreamReader/bs_reader_handle.h:
```diff
@@ -0,0 +1,49 @@
+
+class RocBitstreamReaderHandle {
+public:
+    explicit RocBitstreamReaderHandle(char *input_file_path) { CreateBitstreamReader(input_file_path); };
```
Instead of calling an additional `CreateBitstreamReader` function here, you can use the member initialization list to directly initialize `bs_reader_`. -> 

`explicit RocBitstreamReaderHandle(char *input_file_path) : bs_reader_(std::make_shared<RocVideoESParser>(input_file_path)) {};`

jeffqjiangNew replied:
Done.

LakshmiKumar23 commented on samples/videoDecodeBasic/CMakeLists.txt:
```diff
@@ -0,0 +1,81 @@
+################################################################################
```
Is this going to be a new sample? The `videoDecodeBasic` vs `videoDecode` names could be confusing

jeffqjiangNew replied:
This is a new sample. The reason for the "Basic" suffix is to tell it from the more general app videoDecode. Basic means it doesn't depend on FFMPEG libs.

LakshmiKumar23 commented on samples/videoDecodeBasic/CMakeLists.txt:
```diff
@@ -0,0 +1,81 @@
+
+list(APPEND CMAKE_MODULE_PATH ${PROJECT_SOURCE_DIR}/../../cmake)
+list(APPEND CMAKE_PREFIX_PATH ${ROCM_PATH}/hip ${ROCM_PATH})
+set(CMAKE_CXX_COMPILER ${ROCM_PATH}/bin/amdclang++)
```
Please add a check around this like in this PR: https://github.com/ROCm/rocDecode/pull/466

jeffqjiangNew replied:
Done.

LakshmiKumar23 commented on samples/videoDecodeBasic/README.md:
```diff
@@ -0,0 +1,29 @@
+# Video decode sample
+
+The video decode baisc sample illustrates decoding a single packetized video stream using the built-in bitstream reader, video parser, and rocDecoder to get the individual decoded frames in YUV format. This sample can be configured with a device ID and optionally able to dump the output to a file. This sample uses the high-level RocVideoDecoder class which connects both the video parser and Rocde ...
```
spell check for `basic`

jeffqjiangNew replied:
Done.

LakshmiKumar23 commented on samples/videoDecodeBasic/README.md:
````diff
@@ -0,0 +1,29 @@
+## Run
+
+```shell
+./videodecode -i <input video file [required]> 
````
this would change to `./videodecodebasic` -- based on your executable in CMakeLists

jeffqjiangNew replied:
Done.

LakshmiKumar23 commented on samples/videoDecodeBasic/README.md:
````diff
@@ -0,0 +1,29 @@
+## Build
+
+```shell
+mkdir video_decode_sample && cd video_decode_sample
````
can we call this `mkdir video_decode_basic_sample` for consistency with other sample READMEs

jeffqjiangNew replied:
Done.

LakshmiKumar23 commented on samples/videoDecodePerf/CMakeLists.txt:
```diff
@@ -68,14 +68,14 @@ if(HIP_FOUND AND FFMPEG_FOUND AND ROCDECODE_FOUND AND Threads_FOUND)
     set(LINK_LIBRARY_LIST ${LINK_LIBRARY_LIST} ${FFMPEG_LIBRARIES})
     # rocDecode and utils
-    include_directories (${ROCDECODE_INCLUDE_DIR} ${CMAKE_CURRENT_SOURCE_DIR}/../../utils ${CMAKE_CURRENT_SOURCE_DIR}/../../utils/rocvideodecode ${CMAKE_CURRENT_SOURCE_DIR}/..)
+    include_directories (${ROCDECODE_INCLUDE_DIR} ${CMAKE_CURRENT_SOURCE_DIR}/../../utils ${CMAKE_CURRENT_SOURCE_DIR}/../../utils/rocvideodecode ${CMAKE_CURRENT_SOURCE_DIR}/../../utils/rocmd5 ${CMAKE_CURRENT_SOURCE_DIR}/..)
```
do we need this linked for the Perf sample if we do not use it?

jeffqjiangNew replied:
Yes in order to build because common.h needs MD5. I plan to remove reference to common.h in perf sample in the following PR.

jeffqjiangNew committed:
* rocDecode/Bitstream reader: Added changes based on review comments.

jeffqjiangNew committed:
* rocDecode/Bitstream reader: File name changes based on review comments.

jeffqjiangNew committed:
Merge branch 'develop' into jj/es_file_parser

rrawther commented on utils/rocmd5/roc_md5.cpp:
```diff
@@ -0,0 +1,127 @@
+THE SOFTWARE.
+*/
+
+#include "roc_md5.h"
```
I don't think we need a separate folder for these. I would create a single file e.g md5_utils.h and include both declaration and definition in the same file. Just like video_demuxer.h

jeffqjiangNew replied:
I was thinking to put in our own MD5 implementation in the future so the folder. I'll change it to a header file for now.

jeffqjiangNew replied:
Done.

rrawther commented on samples/videoDecode/CMakeLists.txt:
```diff
@@ -68,10 +68,10 @@ if(HIP_FOUND AND FFMPEG_FOUND AND ROCDECODE_FOUND)
                         ${AVFORMAT_INCLUDE_DIR})
     set(LINK_LIBRARY_LIST ${LINK_LIBRARY_LIST} ${FFMPEG_LIBRARIES})
     # rocDecode and utils
-    include_directories (${ROCDECODE_INCLUDE_DIR} ${CMAKE_CURRENT_SOURCE_DIR}/.. ${CMAKE_CURRENT_SOURCE_DIR}/../../utils ${CMAKE_CURRENT_SOURCE_DIR}/../../utils/rocvideodecode)
```
please remove roc_md5 folder and have one include file instead in utils

jeffqjiangNew replied:
Done.

rrawther commented on samples/videoDecode/README.md:
```diff
@@ -38,4 +38,5 @@ make -j
               -md5_check MD5_File_Path <generate MD5 message digest on the decoded YUV image sequence and compare to the reference MD5 string in a file [optional]>
               -crop <crop rectangle for output (not used when using interopped decoded frame) [optional - default: 0,0,0,0]>
               -m <output_surface_memory_type - decoded surface memory [optional - default: 0][0 : OUT_SURFACE_MEM_DEV_INTERNAL/ 1 : OUT_SURFACE_MEM_DEV_COPIED/ 2 : OUT_SURFACE_MEM_HOST_COPIED/3 : OUT_SURFACE_MEM_NOT_MAPPED]>
+              -no_ffmpeg_demux <use the built-in bitstream reader instead of FFMPEG demuxer to obtain picture data [optional]>
```
does this option handle all the streams. Otherwise, we need not add this option here as we have a separate sample

jeffqjiangNew replied:
We need to add the bitstream reader option to the general decode sample to use MD5, for conformance tests. The videodecoderaw sample does not use any FFMPEG utilities.

jeffqjiangNew replied:
Done.

rrawther commented on samples/videoDecodeBatch/CMakeLists.txt:
```diff
@@ -72,14 +72,14 @@ if(HIP_FOUND AND FFMPEG_FOUND AND ROCDECODE_FOUND)
     endif()
     # rocDecode and utils
-    include_directories (${ROCDECODE_INCLUDE_DIR} ${CMAKE_CURRENT_SOURCE_DIR}/.. ${CMAKE_CURRENT_SOURCE_DIR}/../../utils ${CMAKE_CURRENT_SOURCE_DIR}/../../utils/rocvideodecode)
+    include_directories (${ROCDECODE_INCLUDE_DIR} ${CMAKE_CURRENT_SOURCE_DIR}/.. ${CMAKE_CURRENT_SOURCE_DIR}/../../utils ${CMAKE_CURRENT_SOURCE_DIR}/../../utils/rocvideodecode ${CMAKE_CURRENT_SOURCE_DIR}/../../utils/rocmd5)
```
please remove roc_md5 folder and add a single header file

jeffqjiangNew replied:
Done.

rrawther commented on samples/videoDecodeMem/CMakeLists.txt:
```diff
@@ -68,10 +68,10 @@ if(HIP_FOUND AND FFMPEG_FOUND AND ROCDECODE_FOUND)
     set(LINK_LIBRARY_LIST ${LINK_LIBRARY_LIST} ${FFMPEG_LIBRARIES})
     # rocDecode and utils
-    include_directories (${ROCDECODE_INCLUDE_DIR} ${CMAKE_CURRENT_SOURCE_DIR}/../../utils ${CMAKE_CURRENT_SOURCE_DIR}/../../utils/rocvideodecode)
+    include_directories (${ROCDECODE_INCLUDE_DIR} ${CMAKE_CURRENT_SOURCE_DIR}/../../utils ${CMAKE_CURRENT_SOURCE_DIR}/../../utils/rocvideodecode ${CMAKE_CURRENT_SOURCE_DIR}/../../utils/rocmd5)
```
same comment as before

jeffqjiangNew replied:
Done.

rrawther commented on samples/videoDecodeMultiFiles/CMakeLists.txt:
```diff
@@ -68,10 +68,10 @@ if(HIP_FOUND AND FFMPEG_FOUND AND ROCDECODE_FOUND)
     set(LINK_LIBRARY_LIST ${LINK_LIBRARY_LIST} ${FFMPEG_LIBRARIES})
     # rocDecode and utils
-    include_directories (${ROCDECODE_INCLUDE_DIR} ${CMAKE_CURRENT_SOURCE_DIR}/.. ${CMAKE_CURRENT_SOURCE_DIR}/../../utils ${CMAKE_CURRENT_SOURCE_DIR}/../../utils/rocvideodecode)
+    include_directories (${ROCDECODE_INCLUDE_DIR} ${CMAKE_CURRENT_SOURCE_DIR}/.. ${CMAKE_CURRENT_SOURCE_DIR}/../../utils ${CMAKE_CURRENT_SOURCE_DIR}/../../utils/rocvideodecode ${CMAKE_CURRENT_SOURCE_DIR}/../../utils/rocmd5)
```
remove roc_md5 folder

jeffqjiangNew replied:
Done.

rrawther commented on samples/videoDecodeRGB/CMakeLists.txt:
```diff
@@ -89,7 +89,7 @@ if(HIP_FOUND AND FFMPEG_FOUND AND ROCDECODE_FOUND)
     set(LINK_LIBRARY_LIST ${LINK_LIBRARY_LIST} ${FFMPEG_LIBRARIES})
     # rocDecode and utils
-    include_directories (${CMAKE_CURRENT_SOURCE_DIR}/../../utils ${ROCDECODE_INCLUDE_DIR} ${CMAKE_CURRENT_SOURCE_DIR}/../../utils/rocvideodecode)
+    include_directories (${CMAKE_CURRENT_SOURCE_DIR}/../../utils ${ROCDECODE_INCLUDE_DIR} ${CMAKE_CURRENT_SOURCE_DIR}/../../utils/rocvideodecode ${ROCDECODE_INCLUDE_DIR} ${CMAKE_CURRENT_SOURCE_DIR}/../../utils/rocmd5)
```
remove roc_md5 folder as mentioned before

jeffqjiangNew replied:
Done.

rrawther commented on samples/videoToSequence/CMakeLists.txt:
```diff
@@ -68,10 +68,10 @@ if(HIP_FOUND AND FFMPEG_FOUND AND ROCDECODE_FOUND)
     set(LINK_LIBRARY_LIST ${LINK_LIBRARY_LIST} ${FFMPEG_LIBRARIES})
     # rocDecode and utils
-    include_directories (${ROCDECODE_INCLUDE_DIR} ${CMAKE_CURRENT_SOURCE_DIR}/.. ${CMAKE_CURRENT_SOURCE_DIR}/../../utils ${CMAKE_CURRENT_SOURCE_DIR}/../../utils/rocvideodecode)
+    include_directories (${ROCDECODE_INCLUDE_DIR} ${CMAKE_CURRENT_SOURCE_DIR}/.. ${CMAKE_CURRENT_SOURCE_DIR}/../../utils ${CMAKE_CURRENT_SOURCE_DIR}/../../utils/rocvideodecode ${CMAKE_CURRENT_SOURCE_DIR}/../../utils/rocmd5)
```
remove roc_md5

jeffqjiangNew replied:
Done.

rrawther commented on src/bit_stream_reader/bs_reader_handle.h:
```diff
@@ -0,0 +1,45 @@
+
+class RocBitstreamReaderHandle {
+public:
+    explicit RocBitstreamReaderHandle(char *input_file_path) : bs_reader_(std::make_shared<RocVideoESParser>(input_file_path)) {};
```
might be better to change it to "const uint8_t *" as it is readonly parameter. Also it is better to use uint8_t instead of char

jeffqjiangNew replied:
Done.

rrawther commented on src/bit_stream_reader/es_reader.cpp:
```diff
@@ -0,0 +1,1163 @@
+    // Search for the next start code
+    while (!end_of_stream_) {
+        for (i = 0; i < 3; i++) {
+            if (GetByte(curr_byte_offset_ + i, three_bytes + i) == false) {
```
not sure about the logic here

jeffqjiangNew replied:
We need to break out of the entire while loop when there is no more byte to read.

jeffqjiangNew replied:
This is for EOF handling.

rrawther commented on src/bit_stream_reader/es_reader.cpp:
```diff
@@ -0,0 +1,1163 @@
+        }
+        curr_byte_offset_ = (curr_byte_offset_ + 1) % BS_RING_SIZE;
+    }
+    if (num_start_code_ == 0) {
```
we can combine the if else to a single statement return (num_start_code_ != 0);

jeffqjiangNew replied:
Done.

rrawther commented on src/bit_stream_reader/es_reader.cpp:
```diff
@@ -0,0 +1,1163 @@
+    int nal_start, nal_end_plus_1;
+    int nal_size;
+    nal_start = curr_start_code_offset_;
+    if (curr_start_code_offset_ != next_start_code_offset_) {
```
? : statement better here

jeffqjiangNew replied:
Done.

rrawther commented on src/bit_stream_reader/es_reader.cpp:
```diff
@@ -0,0 +1,1163 @@
+        if (ivf_version != 0) {
+            ERR("Stream file error: Incorrect IVF version (" + TOSTR(ivf_version) + "). Should be 0.");
+        }
+        ptr += 2;
```
better to combine below 2 statements to ptr += 4 and add comment accordingly

jeffqjiangNew replied:
Done.

jeffqjiangNew committed:
Merge branch 'develop' into jj/es_file_parser

jeffqjiangNew committed:
* rocDecode/Bitstream reader: Moved MD5 code into single header file. Added changes based on review comments.

jeffqjiangNew committed:
* rocDecode/Bitstream reader: Removed redundant path.

AryanSalmanpour reviewed (changes requested):
@jeffqjiangNew  Please update the version in the CMakeLists [here](https://github.com/ROCm/rocDecode/blob/develop/CMakeLists.txt#L41) and [here](https://github.com/ROCm/rocDecode/blob/develop/api/rocdecode_version.h#L38) too. Please also update the[ CHANGELOG.md](https://github.com/ROCm/rocDecode/blob/develop/CHANGELOG.md). Please add an "Added" section above the "Changed" section. In the "Added" section, include a line that describes the new APIs introduced in this PR.

AryanSalmanpour commented on src/bit_stream_reader/es_reader.cpp:
```diff
@@ -0,0 +1,1149 @@
+    }
+    end_of_file_ = false;
+    end_of_stream_ = false;
+    bs_ring_ = static_cast<uint8_t*>(malloc(BS_RING_SIZE));
```
It is recommended to use C++ `new` and `delete` for memory allocation and deallocation instead of using `malloc` and `free` functions.

Why is dynamic allocation needed for `bs_ring_` if its size is always `BS_RING_SIZE`? Could you use a fixed array for `bs_ring_` instead?

jeffqjiangNew replied:
Done.

jeffqjiangNew commented:
> @jeffqjiangNew Please update the version in the CMakeLists [here](https://github.com/ROCm/rocDecode/blob/develop/CMakeLists.txt#L41) and [here](https://github.com/ROCm/rocDecode/blob/develop/api/rocdecode_version.h#L38) too. Please also update the[ CHANGELOG.md](https://github.com/ROCm/rocDecode/blob/develop/CHANGELOG.md). Please add an "Added" section above the "Changed" section. In the "Added" section, include a line that describes the new APIs introduced in this PR.

I was going to do that in the following PR. Sure will do here.

jeffqjiangNew committed:
Merge branch 'develop' into jj/es_file_parser

jeffqjiangNew committed:
* rocDecode/Bitstream reader: Changed rocDecode version to 0.10.0. Added minor changes based on review comments.

jeffqjiangNew commented:
> > @jeffqjiangNew Please update the version in the CMakeLists [here](https://github.com/ROCm/rocDecode/blob/develop/CMakeLists.txt#L41) and [here](https://github.com/ROCm/rocDecode/blob/develop/api/rocdecode_version.h#L38) too. Please also update the[ CHANGELOG.md](https://github.com/ROCm/rocDecode/blob/develop/CHANGELOG.md). Please add an "Added" section above the "Changed" section. In the "Added" section, include a line that describes the new APIs introduced in this PR.
> 
> I was going to do that in the following PR. Sure will do here.

Done.
